operator/watchers/k8s_service_sync.go

```diff
@@ -132,21 +141,69 @@ func StartSynchronizingServices(shared bool, cfg ServiceSyncConfiguration) {
 	}()
 
 	swgSvcs := lock.NewStoppableWaitGroup()
-
 	swgEps := lock.NewStoppableWaitGroup()
 
-	// Watch for v1.Service changes and push changes into ServiceCache
-	_, svcController := informer.NewInformer(
+	serviceCacheCallback := newServiceCacheCallback(swgSvcs, swgEps)
+	serviceCallbacks.Register(serviceCacheCallback)
+
+	InitServiceWatcher(cfg, swgSvcs, swgEps, serviceOptsModifier)
+
+	var (
+		endpointController cache.Controller
+	)
+
+	// We only enable either "Endpoints" or "EndpointSlice"
+	switch {
+	case k8s.SupportsEndpointSlice():
+		var endpointSliceEnabled bool
+		endpointController, endpointSliceEnabled = endpointSlicesInit(k8s.WatcherClient(), swgEps)
+		// the cluster has endpoint slices so we should not check for v1.Endpoints
+		if endpointSliceEnabled {
+			break
+		}
+		fallthrough
+	default:
+		endpointController = endpointsInit(k8s.WatcherClient(), swgEps, serviceOptsModifier)
+		go endpointController.Run(wait.NeverStop)
+	}
+
+	go func() {
+		<-k8sSvcCacheSynced
+		cache.WaitForCacheSync(wait.NeverStop, endpointController.HasSynced)
+	}()
+
+	go func() {
+		<-readyChan
+		log.Info("Starting to synchronize Kubernetes services to kvstore")
+		k8sServiceHandler(cfg.LocalClusterName())
+	}()
+}
+
+// InitServiceWatcher creates and runs the v1.Service watcher which watches for
+// changes and push changes into ServiceCache.
+func InitServiceWatcher(
+	cfg ServiceSyncConfiguration,
+	swgSvcs, swgEps *lock.StoppableWaitGroup,
+	optsModifier func(options *v1meta.ListOptions),
+) {
+	// This ensures we only have one instance of a service controller.
+	if serviceController != nil {
+		return
+	}
+
+	// Watch for v1.Service changes and push changes into ServiceCache.
+	serviceIndexer, serviceController = informer.NewInformer(
 		cache.NewFilteredListWatchFromClient(k8s.WatcherClient().CoreV1().RESTClient(),
-			"services", v1.NamespaceAll, serviceOptsModifier),
+			"services", v1.NamespaceAll, optsModifier),
 		&slim_corev1.Service{},
 		0,
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: func(obj interface{}) {
 				metrics.EventTSK8s.SetToCurrentTime()
 				if k8sSvc := k8s.ObjToV1Services(obj); k8sSvc != nil {
-					log.WithField(logfields.ServiceName, k8sSvc.Name).Debugf("Received service addition %+v", k8sSvc)
-					K8sSvcCache.UpdateService(k8sSvc, swgSvcs)
+					for _, f := range serviceCallbacks.Callbacks() {
+						f.OnAdd(k8sSvc)
+					}
```

Comment:
To avoid doing a copy of callbacks every time we receive an event consider executing a function in `serviceCallbacks` that iterates over all callbacks. Something like a). Or consider `panic` if the developer registers another callback after the first callback function call. We do a similar thing like this in other places @brb might help since he is an author of some of that code as I can't remember where it is at this moment.

a)
```golang
func (l *ServiceList) AddAllCallbacks(svc *slim_corev1.Service) {
	for _, cb := range l.cbs {
		cb.OnAdd(svc)
	}
}
```
```golang
				if k8sSvc := k8s.ObjToV1Services(obj); k8sSvc != nil {
					serviceCallbacks.OnAdd(k8sSvc)
				}
```

Reply:
@aanm Which code?

Reply:
Implemented ability to notify call subscribers without creating copies.